CC-1193: Refactored with a new dialect framework for DBMS-specific behavior

rhauch opened the pull request:
The source and sink JDBC connectors were refactored to use dialects for all interaction with the database, allowing for the customization of behavior when working with specific DBMSes. Dialects are extensible and auto-discovered via Java’s service loader mechanism. Separate dialects were added for each major DBMS, though only DBMS-specific behavior already in the connector was extracted into those dialects; additional customization can be done on the dialects to handle DBMS-specific issues.

Prior to this change, the sink connector made use of an earlier dialect framework that was not very extensible, and it only encapsulated a limited subset of the interaction with the database. This new dialect framework is completely extensible (dialects can be added at runtime by simply including in the plugin or on the classpath one or more JARs with the custom dialects), and it completely encapsulates all interaction with the database so that dialects can customize any of the behavior.

A generic dialect is included as a complete implementation that uses JDBC and standard SQL, and custom dialects should extend this generic implementation so that the dialect overrides only the behavior that needs to be customized. The majority of the code within a dialect is needed by the source and sink logic, so there is no longer a distinction between source and sink dialects; each dialect supports all features of both source and sink connectors.

The old sink dialect framework was removed (rather than just deprecated), since it was difficult for users to provide their own custom sink dialect implementations, as the method that determined which sink dialect to use was static and thus not easily customized. Also refactored the sink’s database metadata queries and utilities to reuse components already used in the source.
  
Fixes #303
Fixes #421 
Fixes #35

rhauch commented:
Rebased and corrected the checkstyle errors.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialects.java:
```diff
@@ -0,0 +1,198 @@
+  public static DatabaseDialect findBestFor(
+      String jdbcUrl,
+      AbstractConfig config
+  ) throws ConnectException {
```
This class and in particular this method is the starting point for using dialects. Given a JDBC URL and a source or sink connector configuration, this method iterates through all of the registered `DatabaseDialectProvider` instances and has each provider score the given JDBC URL. The provider with the highest score will be used as a factory to create a new `DatabaseDialect` instance with the supplied connector configuration.

Although `DatabaseDialectProvider` implementations can be registered manually, this class uses Java's Service Provider mechanism to automatically find and register all `DatabaseDialectProvider` implementations that are accessible via this class' classloader. That means that anyone can implement a custom `DatabaseDialectProvider`, build a properly-constructed JAR file w/ provider manifest file, and add that JAR file to the JDBC connector plugin directory. (See the JavaDoc for the `DatabaseDialectProvider` interface for specifics.)

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialectProvider.java:
```diff
@@ -0,0 +1,215 @@
+   * if any of the supplied JDBC subprotocols match the subprotocol from the JDBC connection URL, or
+   * {@link DatabaseDialectProvider#NO_MATCH_SCORE} otherwise.
+   */
+  public abstract static class SubprotocolBasedProvider extends DatabaseDialectProvider {
```
Here's an abstract subclass of `DatabaseDialectProvider` that we used in all of the DBMS-specific dialect providers. Simply provide the subprotocols, and URLs that use one of those subprotocols will be scored with a "perfect" score; anything else gets a "no match" score.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/Db2DatabaseDialect.java:
```diff
@@ -0,0 +1,162 @@
+    public DatabaseDialect create(AbstractConfig config) {
+      return new Db2DatabaseDialect(config);
+    }
+  }
```
Here's an example of how a dialect typically defines a custom `DatabaseDialectProvider` implementation class. It extends the `SubprotocolBasedProvider` so that it perfectly matches any JDBC URL with the `db2`, `db2j`, or `imdb` subprotocols, and it provides a factory method for the actual `Db2DatabaseDialect` implementation class.

Once again, this class is never directly used or referenced anywhere. Instead, the fully-qualified name of this provider class is specified in the `META-INF/services/io.confluent.connect.jdbc.dialect.DatabaseDialectProvider` file

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/Db2DatabaseDialect.java:
```diff
@@ -0,0 +1,162 @@
+   * @param config the connector configuration; may not be null
+   */
+  public Db2DatabaseDialect(AbstractConfig config) {
+    super(config, new IdentifierRules(".", "\"", "\""));
```
The `GenericDatabaseDialect` serves as a base class that implements everything necessary for the JDBC source connector and most methods necessary for the JDBC sink connector. Thus, dialect implementations only implement the methods that are necessary to customize the behavior of the `GenericDatabaseDialect` class.

One of the things the `GenericDatabaseDialect` class does is use the supplied "identifier rules" that dictate the segment delimiter string, leading quote string, and trailing quote string used in the DBMS' fully-qualified identifiers. DB2 uses the period (`.`) for delimiters, and double quotes for leading and trailing quotes. With these rules, the `GenericDatabaseDialect` can build fully-qualified table names used in SQL statements as well as parse strings into fully-qualified names, allowing the connector user to now support either simple table names and fully-qualified names in the whitelist and blacklist filters; if needed, the names can optionally be quoted using the DBMS-specific quoting characters.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialect.java:
```diff
@@ -0,0 +1,471 @@
+ * the connector to find and register your dialect implementation classes, check that your JARs have
+ * the service provider file and your JAR is included in the JDBC connector's plugin directory.
+ */
+public interface DatabaseDialect extends ConnectionProvider {
```
This interface defines methods for all interaction with a database via the JDBC driver. Unfortunately, there are a lot of methods. Some methods may seem strange because they could be implemented by directly using the JDBC driver, but there are lots of different driver implementations and unfortunately all of them have their own quirks.

When you review, please look at the signatures of the public methods, since these will become part of the public API.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialect.java:
```diff
@@ -0,0 +1,471 @@
+   */
+  List<TableId> tableNames(
+      Connection conn
+  ) throws SQLException;
```
One of the challenges dealing with many different JDBC drivers is that they use different combinations of "catalog", "schema", and "table" in their fully-qualified names. To properly deal with these, you can't simply deal with table names as we were before (there are quite a few issues related to simple names not uniquely identifying tables within a database), so I've introduced `TableId` and `ColumnId` to represent fully-qualified table names and column names, and they can deal with any combination of catalog, schema, and table names.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialect.java:
```diff
@@ -0,0 +1,471 @@
+   * @see #identifierRules()
+   * @see IdentifierRules#expressionBuilder()
+   */
+  ExpressionBuilder expressionBuilder();
```
There is quite a bit of code that builds SQL query and update expressions, and all of those need to be able to deal with fully-qualified identifiers _and_ the unique quoting and identifier rules. The `ExpressionBuilder` makes it easier to build a SQL expression without having to worry about those details, since the expression builder delegates to an `IdentifierRules` for the details. However, a DBMS can customize the behavior of `ExpressionBuilder`, and so the dialect must be able to create instances.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialect.java:
```diff
@@ -0,0 +1,471 @@
+  TimestampIncrementingCriteria criteriaFor(
+      ColumnId incrementingColumn,
+      List<ColumnId> timestampColumns
+  );
```
The timestamp and incrementing criteria is some of the most widely varied we've seen across different DBMSes, so this method allows a dialect to generate this criteria. Other similar methods do the same thing for other kinds of criteria and statements.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialect.java:
```diff
@@ -0,0 +1,471 @@
+      int index,
+      Schema schema,
+      Object value
+  ) throws SQLException;
```
This is also something that should be straightforward with the JDBC API, but isn't. Any time a custom dialect overrides the default data type mapping, it may need to change how the values are bound to variables in a prepared statement. For example, `INT8` corresponds to SQL's `BYTE` type, except that SQL `BYTE` is unsigned while `INT8` and Java's `byte` are signed. Most JDBC drivers handle this correctly, but the Sybase driver does nothing special and thus some valid Java `byte` values cannot be inserted into a Sybase `BYTE` column. Therefore, the `SybaseDatabaseDialect` has to instead use `short` to properly handle the signed values, and this changes how `INT8` values are bound to prepared statements for insert/update.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1753 @@
+ * how a column value is converted to a field value for use in a {@link Struct}. To also change the
+ * field's type or schema, also override the {@link #addFieldToSchema} method.
+ */
+public class GenericDatabaseDialect implements DatabaseDialect {
```
The `GenericDatabaseDialect` provides a complete implementation of all methods used by the JDBC source connector and nearly all the methods used by the JDBC sink connector. That does mean that in order to use the JDBC sink connector with a specific DBMS, there must be a custom dialect that _at least_ defines the mapping between Connect's Schema types and the DBMS data types.

This class has to provide a lot of methods and behaviors, and unfortunately that means that this class is quite large. Also, this class needs to be easily extended to override custom behavior, so this class defines a lot of protected methods so that subclasses can override the narrower protected method without having to duplicate a lot of logic in one of the public methods. 

Pulling all of these methods and behaviors into this class makes it very straightforward to implement a custom dialect for a DBMS.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1753 @@
+    public DatabaseDialect create(AbstractConfig config) {
+      return new GenericDatabaseDialect(config);
+    }
+  }
```
The `GenericDatabaseProvider` is one of the few of our dialect implementations that uses the `FixedScoreProvider` base class. The `GenericDatabaseProvider` is the _default_ provider, so it always returns the "average" score, which is lower than a "perfect" score (when the URL has a matching subprotocol supported by the dialect) but higher than a "no match" score (when the URL does not have a matching subprotocol).

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1753 @@
+    if (dbPassword != null) {
+      properties.setProperty("password", dbPassword.value());
+    }
+    properties = addConnectionProperties(properties);
```
The `addConnectionProperties(Properties)` method is an example of a protected method that allows subclasses to override which properties are used without having to duplicate all of the non-trivial logic in this method. Method like this make it easier to subclass without duplicating a lot of logic, but it makes the class bigger. IMO, small price to pay for smaller subclasses.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1753 @@
+      return connection.isValid(timeout);
+    }
+    // issue a test query ...
+    String query = checkConnectionQuery();
```
Here's an example of how we can't simply call a JDBC 4 method, since they won't work on drivers that are only at JDBC 2 or 3.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1753 @@
+                                searchable != null ? searchable.booleanValue() : false,
+                                currency != null ? currency.booleanValue() : false,
+                                isPrimaryKey != null ? isPrimaryKey.booleanValue() : false
+    );
```
This is another example of a method that can be easily overridden to customize the behavior, without having to override or copy the bulk of the logic for querying this information from the database.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1753 @@
+            return rs.getBoolean(col);
+          }
+        };
+      }
```
Okay, I debated about going this route of returning a function `ColumnConverter` that can convert a column value into an object used in a source record, versus just having this method do the conversion and return the value directly. Dealing with functions in Java 7 is verbose, but in Java 8 nearly all of this boilerplate will go away. For example, the above block in Java 8 can be:

```
case Types.BOOLEAN: {
  return (rs)->rs.getBoolean(col);
}
```
Now, the benefit of returning a function is that any overhead of looking at the column definition or whether the driver supports JDBC 4 can be done once, and the resulting function can be called many times (once per record) without this overhead.

So, I decided that returning a function is (or rather _can be_) significantly more efficient in some cases, and it's worth the short-term verbosity for the next 6 months.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1753 @@
+                public Object convert(ResultSet rs) throws SQLException {
+                  return rs.getLong(col);
+                }
+              };
```
While most of these cases don't have any custom logic, here's an example of where the overhead of looking at the precision and scale for a column becomes too expensive to do for every field in every row, and where returning a function is far superior.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1753 @@
+    builder.append(") VALUES(");
+    builder.appendMultiple(",", "?", keyColumns.size() + nonKeyColumns.size());
+    builder.append(")");
+    return builder.toString();
```
Here's an example of how the `ExpressionBuilder` is used. Here, `table` is a `TableId` and can be appended to the expression builder and automatically expanded into the fully-qualified quoted identifier. Line 1405 appends each key column and non-key column but delimited by a comma and properly quoted names (rather than fully-qualified identifiers).

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/MySqlDatabaseDialect.java:
```diff
@@ -0,0 +1,136 @@
+  }
+
+  @Override
+  protected String getSqlType(SinkRecordField field) {
```
This is one of the few methods that _must_ be implemented in order to use this dialect in the sink connector. It maps from the record's schema type to the DBMS-specific data type.

But also note how this whole class is actually pretty short!

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/OracleDatabaseDialect.java:
```diff
@@ -0,0 +1,210 @@
+      final TableId table,
+      Collection<ColumnId> keyColumns,
+      Collection<ColumnId> nonKeyColumns
+  ) {
```
Oracle UPSERT statements get nasty...

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/PostgreSqlDatabaseDialect.java:
```diff
@@ -0,0 +1,261 @@
+    }
+
+    // Delegate for the remaining logic
+    return super.createColumnConverter(mapping);
```
Here's an example of a dialect that first handles DBMS-specific data types (e.g., `BIT` and `JSON`/`JSONB`), and then simply delegates to the `GenericDatabaseDialect` method for all of the standard types.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/SybaseDatabaseDialect.java:
```diff
@@ -0,0 +1,290 @@
+      }
+    } finally {
+      connection.setAutoCommit(autoCommit);
+    }
```
Sybase is one of those DBMSes that does everything a bit differently. The JDBC sink connector turns off autocommit, so any DDL statements it executes is followed by an explicit commit. Of course Sybase doesn't like this, so here we have to temporarily turn on autocommit, apply the DDL statements, and then turn back off autocommit.

rhauch commented on src/main/java/io/confluent/connect/jdbc/sink/dialect/DbDialect.java:
```diff
@@ -1,320 +0,0 @@
-import static io.confluent.connect.jdbc.sink.dialect.StringBuilderUtil.joinToBuilder;
-import static io.confluent.connect.jdbc.sink.dialect.StringBuilderUtil.copiesToBuilder;
-
-public abstract class DbDialect {
```
The JDBC sink connector already had a form of dialects. They didn't support much functionality, and still used the static `JdbcUtils` class to do most of the work. And, the sink connector used a static method to determine which dialect to use, which meant it couldn't be customized without subclassing the connector and then duplicating a lot of code. A search of GitHub and Google didn't turn up any subclasses or custom sink dialects.

Although the issue was around adding dialects for the _source connector_, I took this opportunity to refactor the existing dialects and have the new extensible `DatabaseDialect` framework support both sources and sinks. This helped to eliminate some duplicate code, and now a single custom `DatabaseDialect` implementation means that the source and sink connector will work with the DBMS.

rhauch commented on src/main/java/io/confluent/connect/jdbc/source/BulkTableQuerier.java:
```diff
@@ -66,18 +72,29 @@ protected ResultSet executeQuery() throws SQLException {
 
   @Override
   public SourceRecord extractRecord() throws SQLException {
-    Struct record = DataConverter.convertRecord(schema, resultSet, mapNumerics);
```
Static methods like this are so difficult to customize.

rhauch commented on src/main/java/io/confluent/connect/jdbc/source/JdbcSourceConnectorConfig.java:
```diff
@@ -148,12 +149,21 @@
+      + "  * null (default) means that the catalog name should not be used to narrow the search"
+      + " so that all table metadata would be fetched, regardless their catalog.";
+  private static final String CATALOG_PATTERN_DISPLAY = "Schema pattern";
+  public static final String CATALOG_PATTERN_DEFAULT = null;
```
Some DBMSes like Sybase and SQL Server use the catalog for the database and the schema for the owner. That means that to support these DBMSes (and others like them) we have to allow users to be able to specify a catalog pattern.

rhauch commented on src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java:
```diff
@@ -111,12 +121,10 @@ public void start(Map<String, String> properties) {
     }
 
-    String schemaPattern
-        = config.getString(JdbcSourceTaskConfig.SCHEMA_PATTERN_CONFIG);
```
The schema pattern and some of the other config properties are no longer needed by the task. However, they may be needed by the dialect, so if they're needed that's where they're used.

rhauch commented:
@kkonstantine and @wicknicks: this is ready for an in-depth review.

wicknicks commented on src/main/java/io/confluent/connect/jdbc/sink/BufferedRecords.java:
```diff
@@ -26,61 +26,78 @@
   private static final Logger log = LoggerFactory.getLogger(BufferedRecords.class);
 
-  private final String tableName;
+  private final TableId tableId;
```
+1 for replacing `String` identifier to an object one.

wicknicks commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1753 @@
+  @Override
+  public PreparedStatement createPreparedStatement(
+      Connection db,
+      String query
```
should this method be here? It seems like it should be in a JDBC utils package/class than in a dialect implementation.

rhauch replied:
It does seem strange, but yes. This gives dialects the ability to set fetch direction, query timeouts, max rows for the result sets, etc.

rhauch replied:
Also, IMO it does not make sense to use a static JDBC utility method for things like this, since doing so prevents the ability customize the behavior -- that's what this PR is trying to move away from in the first place.

Now, one might argue that simply creating a prepared statement is something so basic that you shouldn't need anything but the JDBC API. And when you're working with a single DBMS that is definitely true. However, we're dealing with _lots_ of driver implementations and DBMSes, and they all have their quirks, limitations, and pain points. IMO, we _have_ to be prepared to customize almost all interaction with the JDBC driver.

wicknicks commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1753 @@
+    if (parts.size() == 3) {
+      return new TableId(parts.get(0), parts.get(1), parts.get(2));
+    }
+    assert parts.size() >= 2;
```
this is always true

rhauch replied:
Correct, which is why I used an assertion. It's a great way to clarify the logic, but since asserts can be turned off there's no harm in stating them.

rhauch commented on src/main/java/io/confluent/connect/jdbc/util/IdentifierRules.java:
```diff
@@ -0,0 +1,132 @@
+    return new ExpressionBuilder(this);
+  }
+
+  public List<String> parseQualifiedIdentifier(String fqn) {
```
Not easily if at all. We have to be able to look for opening and closing quotes, since the valid characters are different inside quotes. Second, the variable nature of the actual char sequences we're looking for changes, and that'd make the regex pattern string more complicated to understand. Third, DBMSes are allowed to support multi-character quoting strings, which would complicate the regex even more. And finally, regexes aren't super performant and are useful when they would be dramatically easier to code; a really complicated regex loses all those benefits.

kkonstantine commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1753 @@
+
+  @Override
+  public TableId parseTableIdentifier(String fqn) {
+    List<String> parts = identifierRules().parseQualifiedIdentifier(fqn);
```
Knowing how diligent you are with javadocs, it's surprising that `parseQualifiedIdentifier` doesn't have one. Especially given that is returns this very particular list of strings. As a follow up, should this method return a class or an enum (depending on how extensible we expect it to be) instead of a list? I see the convenience behind using a list but it's hard to follow that's expected here. Not the end of the world either.

kkonstantine commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1753 @@
+      Calendar cal
+  ) throws SQLException, ConnectException {
+    String query = currentTimestampDatabaseQuery();
+    assert query != null;
```
nit: Is `assert query != null && !query.isEmpty();` too condensed? I'm used to see those two checks together.

rhauch replied:
If the assertion fails when they are combined, you don't know whether it is null or empty. By having separate assertions, when it does fail you know exactly why.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1753 @@
+      schemaPattern = config.getString(JdbcSourceTaskConfig.SCHEMA_PATTERN_CONFIG);
+      tableTypes = new HashSet<>(config.getList(JdbcSourceTaskConfig.TABLE_TYPE_CONFIG));
+    } else {
+      mapNumerics = JdbcSourceTaskConfig.NUMERIC_PRECISION_MAPPING_DEFAULT;
```
The dialects are used for both sources and sinks, and the `JdbcSourceTaskConfig` properties are only used in methods used by the source task.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1753 @@
+    this.config = config;
+    this.defaultIdentifierRules = defaultIdentifierRules;
+    this.jdbcUrl = config.getString(JdbcSourceConnectorConfig.CONNECTION_URL_CONFIG);
+    if (!(config instanceof JdbcSinkConfig)) {
```
yeah, probably.

MrTrustworthy commented:
Reposting and expanding upon a comment on another thread on request of @rhauch 

I'm not exactly sure how it would be possible to run both a Debezium-MySQL dialect and a "normal" MySQL dialect on the same connect worker since they both have the same JDBC connection string. Maybe an additional config options "sql.dialect.hint" or something would be a useful tool that would allow dialects to break scoring ties in findBestFor()?

Ideally, it would not be necessary to implement an entire dialect to just provide a better debezium data type integration. I've played around with a "dialect extension" approach on the old dialect implementation but decided to scrap that approach due to too many roadblocks. Basically, in order to provide some extension mechanism for dialects that can deal with all eventualities in all database types, you'd need to provide so many hooks it would make the whole process harder than just to write a new dialect. 

But maybe your new implementation would make this feasible. For reference, in order to implement the debezium dialect, I needed to:
(1) have access to the schema from the schema registry whenever a PreparedStatement is assembled
(2) overwrite parts of the strings that are assembled into the queries. For CREATE TABLE statements, i needed to check for debezium datatypes in the schema and pick an appropriate MySQL datatype. For INSERT/UPDATE/UPSERT/... statements I relied on built-in MySQL functions to parse the data (but I'm not sure if this will work for all cases/databases). For example, in order to turn an integer back into a DATETIME for the MySQL, I would write a "FROM_UNIXTIME(?)" instead of a plain "?" in the string that will be used in a PreparedStatement.

Hope that helps
Felix

fcracker79 commented:
Is there any news about merging this PR?

rhauch commented:
@fcracker79, this did not make the cut for the upcoming Confluent Platform release since it was too risky at too late a date. I'm hoping to merge this soon, though.

rhauch commented:
@MrTrustworthy, I'm not sure what you mean by "Debezium dialect". Debezium is an alternative Kafka Connect source connector that does not use JDBC in the same way that this connector does -- it only uses it to obtain the initial snapshot, and once that snapshot is complete it uses MySQL-specific APIs.

Now, you also talk about the Debezium data format. Debezium itself provides some SMTs, and one of those is able to unwrap the row-based information out of Debezium's envelope structure.

Despite that, I do think a dialect hint may be useful in some situations, so that's something we should probably incorporate. (Many drivers that support the same database use different subprotocols, so I'm not convinced that's a problem.)

MrTrustworthy commented:
@rhauch, to clarify what I mean regarding a debezium dialect:

Debezium supplies some additional information about data types in the SMT-unwraped data. For example, a timestamp in the source database might be serialized as INTEGER with the schemaName set to "io.debezium.time.MicroTimestamp". The default MySQL JDBC sink implementation can't deal with that and will not write the expected values into the sink database. The debezium dialect I've written in #367 is basically just a modified MySQL dialect that can deal with this additional meta information and correctly parse it to create both DDL and INSERT statements.

To go back to the dialect hint: Since it's basically just a specialized variation of the MySQL dialect, I needed some way to tell the connector it shouldn't use the standard MySQL dialect without having to change the JDBC connection string.

rhauch commented:
See also #404. The PostgreSQL dialect does not properly handle byte strings, since it uses `BLOB` (which exists and works, but isn't the same as MySQL BLOB, for example) and should instead use `BYTEA`.

rhauch commented on src/main/java/io/confluent/connect/jdbc/JdbcSourceConnector.java:
```diff
@@ -169,8 +166,19 @@ public void stop() throws ConnectException {
+            dialect.close();
+          }
+        } finally {
+          dialect = null;
```
This is more to handle the case if `close()` is called more than once.

rhauch commented on src/main/java/io/confluent/connect/jdbc/JdbcSourceConnector.java:
```diff
@@ -169,8 +166,19 @@ public void stop() throws ConnectException {
+      } finally {
+        try {
+          if (dialect != null) {
+            dialect.close();
```
Removed from the connection provider and kept in this class, since this is the class that instantiates the dialect.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialect.java:
```diff
@@ -0,0 +1,471 @@
+
+/**
+ * A component that is used to map datatypes and queries for JDBC sources. Every dialect should be
+ * threadsafe and all methods should be free of side effects.
```
ack

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialect.java:
```diff
@@ -0,0 +1,471 @@
+   * @throws SQLException if there is an error with the database connection
+   */
+  Timestamp currentTimeOnDB(
+      Connection conn,
```
ack

rhauch commented:
@kkonstantine, @wicknicks: I've rebased and hopefully addressed all of your comments, so this PR is ready for another review.

@MrTrustworthy, I've added a `dialect.name` configuration that when no provided or empty (the default), then the best dialect is used. However, when `dialect.name` is provided and matches the name of one of the available dialects, then that dialect will be used. Note that there is a recommender and validator, so the `dialect.name` configuration will be validated if given. Hopefully this satisfies your requirements by allowing you to explicitly specify which dialect you want to use.

rhauch commented:
retest this please

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1727 @@
+        break;
+      }
+
+      case Types.NUMERIC:
```
@kkonstantine this is the first of two blocks which incorporate the substantiative portions of your recent changes from #396. Would love a sanity check here.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1727 @@
+        return (rs) -> rs.getDouble(col);
+      }
+
+      case Types.NUMERIC:
```
@kkonstantine this is the second of two blocks which incorporate the substantiative portions of your recent changes from #396. Would love a sanity check here.

rayokota commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1727 @@
+  public void close() {
+    for (Connection conn = connections.poll(); conn != null; conn = connections.poll()) {
+      try {
+        conn.close();
```
Maybe use `org.apache.kafka.common.utils.Utils.closeQuietly`, which will catch `Throwable`?  It seems if `close()` throws a `Throwable`, the other connections will not get cleaned up.  This could cause issues on the server side.

rayokota commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1727 @@
+      try (Statement statement = connection.createStatement()) {
+        boolean hasResultSet = statement.execute(query);
+        if (hasResultSet) {
+          statement.getResultSet().close();
```
This `close()` should be in a finally block; otherwise a leak may occur

rhauch replied:
I just changed it to catch `Throwable` instead of `SQLException`.

rhauch replied:
+1

rhauch replied:
@rayokota, actually, we're not doing anything with the result set other than closing it. What would be in the try block?

rayokota commented on src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java:
```diff
@@ -131,81 +144,102 @@ public void start(Map<String, String> properties) {
+      cachedConnectionProvider = null;
+      try {
+        if (dialect != null) {
+          dialect.close();
```
Should we use `org.apache.kafka.common.utils.Utils.closeQuietly` ?

rhauch replied:
I actually like the warning appearing on the same logger, so I just caught Throwable here and logged a warning.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1727 @@
+      if (defaultIdentifierRules != null && connections.peek() == null) {
+        // We've not yet established a connection to the database, so in this case
+        // just return the default rules ...
+        return defaultIdentifierRules;
```
I was trying to avoid creating a connection just to be able to use the identifier rules in the generic case. But I think you're right -- that behavior is unexpected. All of the concrete types define specific identifier rules, so it's only the generic case that needs to talk to the database. Later in the method, If we can't establish a connection, we log a warning and use the defaults. So, I'll remove this if-block altogether.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialect.java:
```diff
@@ -0,0 +1,461 @@
+   * @param nonKeyColumns the identifiers of the other columns in the table; may not be null but may
+   *                      be empty
+   * @return the upsert/merge statement; may not be null
+   */
```
It's not caught, so using upsert with the general dialect will result in a failure, and this is the same behavior as before. Basically, you can't use upsert without a concrete dialect that overrides it.

I didn't add it to the JavaDoc because I thought it implied that it makes the implementation optional, when it really isn't optional if you want to use upsert.

However, I'll add some logic where this method is used to report a more meaningful exception.

C0urante commented on src/main/java/io/confluent/connect/jdbc/util/ExpressionBuilder.java:
```diff
@@ -0,0 +1,294 @@
+
+import javax.xml.bind.DatatypeConverter;
+
+public class ExpressionBuilder {
```
It'd be nice to have a little documentation on this class and some of its more commonly-used methods, since it's referenced pretty widely across the project.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialectProvider.java:
```diff
@@ -0,0 +1,208 @@
+  public static final int AVERAGE_MATCHING_SCORE = 10;
+
+  /**
+   * The score for a dialect that is an excellent match.
```
Yeah, there is currently no maximum. Do you suggest changing something?

C0urante replied:
Either change the constant name to `EXCELLENT_MATCHING_SCORE` or change the value to `Integer.MAX_VALUE`. I think the first choice is better.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1727 @@
+        }
+      }
+    }
+    return true;
```
Yes, an exception will signify that the connection is not healthy. If the query succeeds, then we consider the connection health.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1727 @@
+   * @return the query string; never null or empty
+   */
+  protected String currentTimestampDatabaseQuery() {
+    return "SELECT CURRENT_TIMESTAMP;";
```
It doesn't matter, but I'll remove the semicolon here.

rhauch commented on src/main/java/io/confluent/connect/jdbc/sink/BufferedRecords.java:
```diff
@@ -169,31 +184,43 @@ public void close() throws SQLException {
     }
   }
+
+  private Collection<ColumnId> asColumns(Collection<String> names) {
```
Agree.

rhauch commented on src/main/java/io/confluent/connect/jdbc/dialect/GenericDatabaseDialect.java:
```diff
@@ -0,0 +1,1727 @@
+        mutability,
+        precision,
+        scale,
+        signedNumbers != null ? signedNumbers.booleanValue() : false,
```
Correct, unboxing is not necessary, except `signedNumber` only makes sense for numbers and thus I'm defining `null` for non-numbers.

C0urante replied:
Not sure I follow. The ternary check here to make sure a default value is used in the event that `signedNumbers` is null makes sense--but when it's non-null, why is an explicit call to `.booleanValue()` necessary? `signedNumbers` will be automatically unboxed from a `Boolean` Object to a `boolean` primitive without the call to `.booleanValue()`.

rhauch commented on src/main/java/io/confluent/connect/jdbc/JdbcSourceConnector.java:
```diff
@@ -169,8 +166,19 @@ public void stop() throws ConnectException {
+      } finally {
+        try {
+          if (dialect != null) {
+            dialect.close();
```
I chose to not bring in the dependency and to simply catch the warning. I like having the log warning message be in the same logger, which will help when configuring logs for the JDBC connector.
